Class1

sayjudith opened the pull request.

sayjudith committed:
changed to classes

sayjudith committed:
fixed folders to correct names, fixed fuctions names

sayjudith committed:
Chanched Struct to Class

sayjudith committed:
Delete Lessons1_4/bin/Debug/net8.0 directory

sayjudith committed:
Changed functions names, deleted this.

sayjudith committed:
Merge branch 'Class1' of https://github.com/sayjudith/Scraping into Class1

sayjudith committed:
renaming functions

sayjudith committed:
deleted rolls and spaces

sayjudith committed:
deleted rolls and spaces

sayjudith committed:
Update NewsModel.cs

slav4ik51493 commented on Lessons1_4/Lessons1_4/Program.cs:
```diff
@@ -1,136 +1,14 @@
-        {
-            return fileName.Replace("\"", "").Replace("\'", "").Replace(":", "").Replace("*", "");
-        }
+        SiteParser spLocal = new SiteParser("https://ria.ru/world/");
```
var

slav4ik51493 commented on Lessons1_4/Lessons1_4/Program.cs:
```diff
@@ -1,136 +1,14 @@
+        SiteParser spLocal = new SiteParser("https://ria.ru/world/");
+        spLocal.GetNewsAsync().Wait();
+        spLocal.PrintListNewsStruct();
+        FileSaver fsLocal = new FileSaver();
```
var

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/FileSaver.cs:
```diff
@@ -0,0 +1,76 @@
+            string currentDirectory = Directory.GetCurrentDirectory();
+            string folderPath = Path.Combine(currentDirectory, FolderName);
+            var i = 0;
+            DirectoryInfo di = new DirectoryInfo(folderPath);
```
var

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,116 @@
+{
+    internal class SiteParser
+    {
+        public string url;
```
Не нужно делать public то, что не используется снаружи. Всё делаем private всегда и по необходимости делаем public.

slav4ik51493 replied:
И если это будет приватным полем, то правило наименования такое, что у тебя должно начинаться с нижнего подчеркивания, аля `_url`. Если несколько слов, то вот пример `_someFieldName`

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,116 @@
+    internal class SiteParser
+    {
+        public string url;
+        public string URL { get; set; }
```
не используется. А если бы использовалось, то имя должно было быть с большой буквы в CamelCase. Аля, `SomePropertyName`.

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,116 @@
+        public string url;
+        public string URL { get; set; }
+
+        public List<NewsModel> News;
```
не используется во вне, поэтому приватным и сделай правильное имя, как описывал выше

sayjudith replied:
List<NewsModel> используется в классе FileSaver

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,116 @@
+
+        public List<NewsModel> News;
+
+        public SiteParser(string ExternalURL)
```
параметры методов/конструкторов всегда пишется в camelCase. Т.е. `externalUrl`. И да, аббревиатуры с 3 и больше буквами должны быть с большой буквы и остальные маленькие. Если 2 буквы в аббревиатуре, то большими примеры:
`Url`, `UI`, `Html`, `Css`. 
Если параметр состоит просто из одной аббревиатуры и является параметров, то просто маленькими буквами, аля:
`string ui`.

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,116 @@
+            return News;
+        }
+
+        public async Task GetNewsAsync()
```
Слишком большой метод, который содержит в себе кучу разной логики парсинга. Давай декомпозируем на разные методы. Аля вот так: 

1. CreateBrowserAsync - содержит в себе создание браузера, страницы
2. GotoAsync можешь оставить без отдельного метода, норм итак, не надо выносить в отдельный метод
3. GetNewsCountAsync()
4. Затем твои проверки и цикл могут жить тут, в нем вызываются следующие методы:
5. ParseNewsDataAsync
6. ParseNewsDetailsAsync
7. GetFullNewsContentAsync

Если я что-то упустил, добавь сама

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,116 @@
+            var playwright = await Playwright.CreateAsync();
+            var browser = await playwright.Chromium.LaunchAsync(new BrowserTypeLaunchOptions { Headless = true });
+            var page = await browser.NewPageAsync();
+            var pageHref = await browser.NewPageAsync();
```
не используется

sayjudith committed:
change types

change types (var)

sayjudith committed:
Update SiteParser.cs

add new methods

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,126 @@
+            }
+        }
+
+        public void PrintListNewsStruct()
```
что тут значит Struct? у нас же нет структур

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,126 @@
+
+        public void PrintListNewsStruct()
+        {
+            foreach (NewsModel anotherNews in News)
```
var вместо NewsModel

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,126 @@
+    internal class SiteParser
+    {
+        private string _url;
+        public string Url { get; set; }
```
у тебя снова _url и Url вместе существуют, зачем?

sayjudith committed:
changed methods names

sayjudith committed:
Update SiteParser.cs

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,156 @@
+                    {
+                        await GetCurrentNewsValuesAsync
+                            (
+                            i, linkReference, headerText, newsPreviewDate, newsPreviewCount, tagsString, pageHref
```
It's not good formatting, it could be do like this:
```
await GetCurrentNewsValuesAsync(i, linkReference, headerText,
    newsPreviewDate, newsPreviewCount, tagsString, pageHref);
```

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,156 @@
+                catch (Exception ex)
+                {
+                    Console.WriteLine("error" + ex.Message + i.ToString());
+                    await GetCurrentNewsValuesAsync(i, string.Empty, string.Empty, string.Empty, string.Empty, string.Empty, pageHref);
```
Этот вариант плохой по логике.
Почему, потому что у тебя проблема была только с newsPreviewDate и newsPreviewCount, это когда вместо нормальных элементов с датой и количеством просмотров мы видели звездочки непонятные.
А ты в текущей логике отказываешься от всех данных.



Их и нужно было обернуть в try-catch, и более того, тут даже можно было бы создать отдельный метод, который бы имел такой примерно вид, смотри:

```
static (string a, string b) GetDateAndViews(ILocator newsLocator)
{
    try
    {
        var newsItemInfos = newsLocator.Nth(i).Locator(".list-item__info-item");
        var newsPreviewDate = await newsItemInfos.Nth(0).TextContentAsync();
        var newsPreviewCount = await newsItemInfos.Nth(1).TextContentAsync();

        return (newsPreviewDate, newsPreviewCount);
    }
    catch(Exception ex)
    {
        Console.WriteLine($"Due getting post date and views, we got exception: {ex.Message}");
        return (string.Empty, string.Empty);
    }
}
```

slav4ik51493 replied:
И да, только что увидел, ты таки получаешь данные о просмотрах и дате на полной странице, и они там вроде как надежно всегда есть.
Тогда к чему мы вообще тут их пытаемся поймать, если мы там точно возьмем? Нам тут звездочки эти жизнь усложняют, предлагаю вообще выпилить это лишнее получение, мы делаем двойную работу = лишний код = доп сложность = доп время для работы скраппера.

slav4ik51493 replied:
Я бы переосмыслил полностью как работает твой GetListOfNewsAsync

Смотри, как тебе нужно сделать, и таки как сделать с одной page, без pageHref. Использование дополнительной страницы в браузере = дополнительные ресурсы. Это плохо влияет на производительность, потому что хром любит жрать оперативу.

Алгоритм такой:
1. запускаешь браузер, одну страницу
2. идешь на главную страницу, получаешь список новостей, и сохраняешь ссылки на них, у тебя уже это в linkReference есть кажется.
3. После того, как ты прошла по всем новостям и собрала по ним основную инфу, в том числе ссылки, теперь ты можешь использовать тот же самый единственный page для последовательной навигации по каждой из новости.
Всё, берешь foreach или for и идешь по List с инфой по твоим новостям и вызываешь в нем метод GetCurrentNewsValuesAsync, который уже пылесосит оставшуюся инфу с каждой новости.
4. После окончания сбора данных, как и есть, ты сохраняешь это в файлы.


Т.е., в итоге, тебе нужно переделать свой метод GetListOfNewsAsync и избавиться от pageHref, он абсолютно точно не нужен, алгоритм описан выше.

sayjudith committed:
Update methods

Changed logic, now we take list with new links and title and etc. than we use page without pageHref

slav4ik51493 commented on Lessons1_4/Lessons1_4/Models/NewsModel.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Lessons1_4.Models
+{
+    internal class NewsTitleModel
```
Мне не очень нравится название. Он вроде бы Title, но при этом содержит всю инфу о себе. 
Давай из двух моделей в этом файле сделаем одну, NewsModel.

slav4ik51493 commented on Lessons1_4/Lessons1_4/Models/NewsModel.cs:
```diff
@@ -0,0 +1,38 @@
+
+    internal class NewsModel
+    {
+        public NewsTitleModel titleModel { get; set; }
```
если таки мы объединим, то удалим это свойство вообще. А так - naming convention здесь был соблюден неверно. 
public свойство именуется с большой буквы в CamelCase! Также, как это сделано в FullContent.

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,174 @@
+            return _news;
+        }
+
+        public async Task GetNewsAsync()
```
При повторном пересмотре кода, заметил, что у нас есть один странный момент.
Метод называется GetNewsAsync и логично, чтобы он возвращал их, а не просто сохранял в _news и трындец.
Поэтому, нам нужно:
1. меняем возвращаемый тип этого метода на Task<List<NewsModel>>.
2. Удаляем ReadNewsFromSite() метод, он рудимент. 
3. внутри текущего (GetNewsAsync) мы сделаем return _news, который заполняется при помощи других методов.

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,174 @@
+                {
+                    Console.WriteLine("error" + ex.Message + i.ToString());
+                    _newsTitles.Add(
+                        new NewsTitleModel() {
```
фигурная скобка на новую строку.
 круглые скобки могут быть также сокращены

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,174 @@
+
+        private async Task GetCurrentNewsValuesAsync(int countOfNews, List<NewsTitleModel> newsTitleList, IPage pageInnerHref)
+        {
+            if (countOfNews > 0) {
```
1. фигурная скобка на новую строку
2. обертывать целый метод в if - не самое удачное решение, это усложняет понимание работы метода. Лучше пойти от обратного, и сначала разобрать негативные сценарии и закончить работу метода при помощи return или выброса исключения (если конечно того требует ситуация). 
Т.е., тебе тут надо сделать так:
```
 private async Task GetCurrentNewsValuesAsync(int countOfNews, List<NewsTitleModel> newsTitleList, IPage pageInnerHref)
        {
            if (countOfNews == 0)
            {
                Console.WriteLine("Wrong list of news, skipping...");
                return;
            }
            // основная логика твоего метода
```

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,174 @@
+            await GetCurrentNewsValuesAsync(newsCount, _newsTitles, pageHref);
+        }
+
+        private async Task GetCurrentNewsValuesAsync(int countOfNews, List<NewsTitleModel> newsTitleList, IPage pageInnerHref)
```
1. countOfNews. Во-первых, название лучше делать без of. newsCount было бы лучше. Во-вторых, у нас есть List, с которым ты как раз и работаешь, он уже имеет .Count и знает количество элементов, не нужно для этого тащить еще один бесполезный вспомогательный параметр в метод.
2. newsTitleList - лучше использовать множественную форму, нежели добавлять постфикс "List". Поэтому лучше назвать newsTitles. Но в связи с тем, что я прошу переименовать название модели, то вовсе получится просто news.
3. pageInnerHref. Не очень понятно почему такое название. Оно больше путает. Здесь у тебя это просто page, она у тебя вообще одна единственная, внутри которой происходит вся работа, других page у нас не открывается. В методе GetListOfNewsAsync я бы тоже изменил название с pageHref -> page.

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,174 @@
+            Console.WriteLine("браузер закрыт\n");
+        }
+
+        private async Task GetListOfNewsAsync(IPage pageHref, ILocator newsLocator, int newsCount) 
```
pageHref -> page.
Объяснение ниже для метода GetCurrentNewsValuesAsync, пункт 3.

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,174 @@
+                    {
+                        var fileName = FormatFileName(headerText);
+                        _newsTitles.Add(
+                            new NewsTitleModel() {
```
фигурная скобка на новую строку.
круглые скобки могут быть также сокращены

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,174 @@
+            if (countOfNews > 0) {
+                for (int i = 0; i < countOfNews; i++)
+                {
+                    var currLinkReference = newsTitleList[i].LinkReference;
```
Сокращать слова подобным образом не очень. Лучше использовать полное, т.е. currentLinkReference.

slav4ik51493 replied:
это все еще читабельно. А если получается veryImportantValueOfAllProgramReallyFromMicroservicesApi (супер тупая белеберда), то стоит лучше подумать над названием переменной, которая будет понятно доносить, зачем она нужна и что в себе хранит

sayjudith committed:
update logic

sayjudith committed:
Update classes

Fix Model to one model, fix methods in other classes

slav4ik51493 commented on Lessons1_4/Lessons1_4/Services/SiteParser.cs:
```diff
@@ -0,0 +1,165 @@
+            await GetCurrentNewsValuesAsync(newsCount, _news, page);
+        }
+
+        private async Task GetCurrentNewsValuesAsync(int newsCount, List<NewsModel> newsTitleList, IPage page)
```
1. вот тут уже newsCount не нужен, потому что у тебя есть List. 
2. И еще, переименуй newsTitleList. Уже тайтлов так сказать у нас нет, это просто должно быть "news".

slav4ik51493 replied:
не забудь убрать еще как сказал в голосовом List<NewsModel> newsTitleList из параметров - он тоже не нужен, потому что у тебя есть _news

sayjudith committed:
Update SiteParser.cs

Update GetCurrenrNewsValueAsync
